packages/SwingSet/src/kernel/virtualObjectManager.js

```diff
@@ -264,86 +333,171 @@ export function makeVirtualObjectManager(
+        // TODO: these should be refcounted instead of this being a journey of no return
+        reachablePresences.add(vref);
+      }
     }
   }
 
-  function isVrefReachable(vref) {
-    return reachableVrefs.has(vref);
+  /**
+   * Check if a given vref points to a reachable presence.
+   *
+   * @param {string} vref  The vref of the presence being enquired about
+   *
+   * @returns {boolean} true if the indicated presence remains reachable.
+   */
+  function isPresenceReachable(vref) {
+    return reachablePresences.has(vref);
   }
 
   /**
-   * Set of all import vrefs which are recognizable by our virtualized data.
-   * These were Presences at one point. We add to this set whenever we use a
-   * Presence as a key into a makeVirtualScalarWeakMap() instance or an instance of
-   * VirtualObjectAwareWeakMap or VirtualObjectAwareWeakSet. We currently never
-   * remove anything from the set, but eventually we'll use refcounts to figure
-   * out when the vref becomes unrecognizable, allowing the vat to send a
-   * retireImport into the kernel.
+   * A map from vrefs (those which are recognizable by (i.e., used as keys in)
+   * VOM aware collections) to sets of recognizers (the collections for which
+   * they are respectively used as keys).  These vrefs correspond to either
+   * imported Presences or virtual objects (Remotables do not participate in
+   * this as they are not keyed by vref but by the actual Remotable objects
+   * themselves). We add to a vref's recognizer set whenever we use a Presence
+   * or virtual object as a key into a makeVirtualScalarWeakMap() instance or an
+   * instance of VirtualObjectAwareWeakMap or VirtualObjectAwareWeakSet.  We
+   * remove it whenever that key (or the whole collection containing it) is
+   * deleted.
+   *
+   * A recognizer is one of:
+   *   Map - the map contained within a VirtualObjectAwareWeakMap to point to its vref-keyed entries.
+   *   Set - the set contained within a VirtualObjectAwareWeakSet to point to its vref-keyed entries.
+   *   deleter - a function within a WeakStore that can be called to remove an entry from that store.
+   *
+   * It is critical that each collection have exactly one recognizer that is
+   * unique to that collection, because the recognizers themselves will be
+   * tracked by their object identities, but the recognizer cannot be the
+   * collection itself else it would prevent the collection from being garbage
+   * collected.
    *
+   * TODO: all the "recognizers" in principle could be, and probably should be,
+   * reduced to deleter functions.  However, since the VirtualObjectAware
+   * collections are actual JavaScript classes I need to take some care to
+   * ensure that I've got the exactly-one-per-collection invariant handled
+   * correctly.
+   *
+   * TODO: concoct a better type def than Set<any>
    */
-  /** @type {Set<string>} of vrefs */
-  const recognizableVrefs = new Set();
+  /** @type {Map<string, Set<any>>} */
```

Comment:
@dckc taught me to never ever use the word `any` in a type. I think this could be expressed as `Set<Map | Set | {}>` instead?